Support type annotations in string frontend

jamesr66a opened the pull request.

Commit:
Add Decl

Commit:
Type parsing via Decl

Commit:
Refactor subscript parsing to support comma-separated lists

Commit:
Parse return type annotations in Decl parsing

Commit:
Parse comment type annotations in string frontend

Commit:
Parse type exprs into FunctionSchema

Commit:
Meaningful tests

Commit:
Test and improve error messages

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1486,13 +1494,96 @@ void defineMethodsInModule(Module & m, const std::vector<TypedDef>& definitions,
+std::unordered_map<std::string, TypePtr> ident_to_type_lut() {
+  static std::unordered_map<std::string, TypePtr> map = {
+    {"Tensor", DynamicType::get()},
+    {"Int", IntType::get()},
```
int

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1486,13 +1494,96 @@ void defineMethodsInModule(Module & m, const std::vector<TypedDef>& definitions,
+  static std::unordered_map<std::string, TypePtr> map = {
+    {"Tensor", DynamicType::get()},
+    {"Int", IntType::get()},
+    {"Float", FloatType::get()},
```
float

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1486,13 +1494,96 @@ void defineMethodsInModule(Module & m, const std::vector<TypedDef>& definitions,
+TypePtr parseTypeFromExpr(Expr expr) {
+  if (expr.kind() == TK_VAR) {
+    auto ident = Var(expr).name();
+    if (ident_to_type_lut().count(ident.name())) {
```
nit: use .find

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1486,13 +1494,96 @@ void defineMethodsInModule(Module & m, const std::vector<TypedDef>& definitions,
+      throw ErrorReport(subscript.value().range()) << "Subscripted type must be a type identifier";
+    }
+    auto value_name = Var(subscript.value()).name().name();
+    if (!subscriptable_types().count(value_name)) {
```
nit: if you are going to just test each valid value_name, there is no need to also have a separate table that has to stay in sync with the if statement chain here.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1486,13 +1494,96 @@ void defineMethodsInModule(Module & m, const std::vector<TypedDef>& definitions,
+                                  << " cannot be used in a type expression";
+}
+
+std::vector<Argument> parseArgsFromDef(Def &def) {
```
These next two should be fromDecl not fromDef, they never use the def.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1486,13 +1494,96 @@ void defineMethodsInModule(Module & m, const std::vector<TypedDef>& definitions,
   }
 }
 
+namespace type_annotations {
```
its a cpp file, we don't really need another namespace

zdevito commented on torch/csrc/jit/script/parser.h:
```diff
@@ -266,6 +283,9 @@ struct Parser {
         auto values = parseList(TK_NOTHING, ',', TK_NEWLINE, &Parser::parseExp);
         return Return::create(range, values);
       }
+      case TK_TYPE_COMMENT: {
```
A type comment can only appear at the beginning of a function so it should not be parsed here.

zdevito commented on torch/csrc/jit/script/parser.h:
```diff
@@ -314,27 +334,79 @@ struct Parser {
+  // First return value is the TreeRef to the list of statements we've parsed.
+  // The second one is an optional Decl for if we find type annotations in a
+  // comments
+  std::tuple<TreeRef, std::shared_ptr<TreeRef>> parseStatements() {
```
This can be significantly cleaned up by changing it to `parseStatement(bool expect_indent=true)`

Then changing parseFunction to:

```
L.expect(TK_INDENT)
if(L.nextIf(TK_TYPE_COMMENT) {
   ....
}
parseStatements(/*expect_indent=*/false)
```
Currently it spends time both enabling type comments to appear where they should be able to appear, and then check that it doesn't happen.

zdevito commented on torch/csrc/jit/script/parser.h:
```diff
@@ -314,27 +334,79 @@ struct Parser {
+    TreeRef stmts_list;
+    std::shared_ptr<TreeRef> type_annotation;
+    std::tie(stmts_list, type_annotation) = parseStatements();
+    if (type_annotation) {
```
Move this to its own function, it obscures the parsing logic of parseFunction

zdevito reviewed (commented):
Overall approach seems good so far. There will still be more changes, I imagine, for parsing the decl when it is in python.  Only comments so far are on places where the code can be simplified.

Commit:
Address comments

Commit:
Fix return type mess

Commit:
Address more comments

Commit:
delete old comment

Commit:
@script -> extract schema from Def instead of TypedDef

Commit:
Exhaustive tests

Commit:
Implement for @script_method

Commit:
Update tests to include method tests for string

Commit:
delete dead code

Commit:
bugfix

Commit:
maybe its not a hack

iotamudelta committed:
enable unit tests and other changes (#10266)

Summary:
This PR for the ROCm target does the following:
* enable some unit tests on ROCm
* fix a missing static_cast that breaks BatchNorm call on ROCm
* fix BatchNorm to work on ROCm w/ ROCm warp sizes etc
* improve the pyhipify script by introducing kernel scope to some transpilations and other improvements
* fix a linking issue on ROCm
* for more unit test sets: mark currently broken tests broken (to be fixed)
* enable THINLTO (phase one) to parallelize linking
* address the first failing of the elementwise kernel by removing non-working ROCm specialization
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10266

Differential Revision: D9184178

Pulled By: ezyang

fbshipit-source-id: 03bcd1fe4ca4dd3241f09634dbd42b6a4c350297

ezyang committed:
Move ScalarType to ATen/core, splitting out Backend

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/10262

Reviewed By: dzhulgakov

Differential Revision: D9157408

fbshipit-source-id: 11631a35dfc6cb1f73f61ea08d3115f8ef4cb034

achalshah20 committed:
Stop propagating std flags to downstream gcc/nvcc (#10098)

Summary:
When we directly use -std=c++11, it propagates to the downstream applications.

Problems:
1. Gcc flags propagating to nvcc.
2. nvcc flags propagating to nvcc. (Which throws an error like redeclaration of std flag)

This PR will fix these propagation issues!

Similar problem:
https://github.com/FloopCZ/tensorflow_cc/pull/92
https://github.com/CGAL/cgal/issues/2775

Requires: Cmake 3.12
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10098

Differential Revision: D9187110

Pulled By: ezyang

fbshipit-source-id: 0e00e6aa3119c77a5b3ea56992ef3bbfecd71d80

Commit:
Fix merge errors
